modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/IRNode.java

```diff
@@ -31,40 +31,56 @@
 
     /* ---- begin decorations ---- */
 
-    public interface IRDecoration {
+    public abstract static class IRDecoration<V> {
```

Comment:
This is a substantional enough class that it should be pulled out into it's own file.

Reply:
I agree, but I'm going to hold off on this for now as I think we should reorg everything into user and ir packages with subpackages for node and phase. I will do that in a follow up PR.